### PR TITLE
feat: add parallel source connector

### PR DESCRIPTION
**Pull request by shanicky**
Signed-off-by: Peng Chen <peng@singularity-data.com>

## What's changed and what's your intention?

This PR attempts to control the parallelization of SplitReader in a generic way

This PR provides a new Source called `ParallelConnectorSource`, which generates a Reader called `ParallelConnectorSourceReader`, which generates multiple `InnerConnectorSourceReader`, each InnerConnectorSourceReader will only hold 1 Split to be used for `SplitReader` consumption.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

File: src/source/src/manager.rs [hunk 1]
```diff
@@ -135,7 +134,7 @@ impl SourceManager for MemSourceManager {
         );
         let row_id_index = info.row_id_index as usize;
 
-        let source = SourceImpl::Connector(ConnectorSource {
+        let source = SourceImpl::ParallelConnector(ParallelConnectorSource {
```
Comment (neverchanje) [inline src/source/src/manager.rs:137]: Then why do `SourceImpl::Connector` still needed? Can we just use `SourceImpl::Connector` to represent parallel connector?

Comment (shanicky) [reply src/source/src/manager.rs]: removed

File: src/source/src/connector_source.rs [hunk 2]
```diff
@@ -93,58 +90,199 @@ impl ConnectorSource {
...
+                biased;
+                // stop chan has high priority
+                _ = stop.borrow_mut() => {
+                    break;
+                }
+
+                chunk = self.reader.next() => {
+                    match chunk.map_err(|e| internal_error(e.to_string())) {
+                        Err(e) => {
+                            // just ignore
```
Comment (tabVersion) [inline src/source/src/connector_source.rs:112]: suggest leaving a log here

**Review by pangzhenzhou — APPROVED** [review]

File: src/source/src/connector_source.rs [hunk 3]
```diff
@@ -93,58 +90,199 @@ impl ConnectorSource {
...
+        splits: Vec<SplitImpl>,
+        column_ids: Vec<ColumnId>,
+    ) -> Result<ConnectorSourceReader> {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let mut handles = HashMap::with_capacity(splits.len());
+        let config = self.config.clone();
+        let columns = self.get_target_columns(column_ids)?;
+
+        let readers = try_join_all(splits.into_iter().map(|split| {
+            log::debug!("spawning pulsar split reader for split {:?}", split);
```
Comment (tabVersion) [inline src/source/src/connector_source.rs:257]: pulsar -> connector

**Review by tabVersion — APPROVED** [review]
rest lgtm

**Pull request merged**